sql,opt: add support for lookup joins with multiple spans

rytaft opened the pull request:
This commit adds support for lookup joins that scan multiple spans in
the index per input row. This is needed for cases when a prefix of the
index columns are constrained to multiple constant values, such as the
values in a `CHECK` constraint or an `ENUM` type.

This is implemented by adding a new field to the `LookupJoinPrivate` in
the optimizer (and the `lookupJoiner` in the execution engine). The new
field is called `LookupExpr`, and similar to `KeyCols` in the `LookupJoinPrivate`
(`lookupCols` in the `lookupJoiner`), `LookupExpr` represents the part of the
join condition used to perform the lookup into the index. It should only
be set when `KeyCols` is empty. `LookupExpr` is used instead of `KeyCols` when
the lookup condition is more complicated than a simple equality between
input columns and index columns. In this case, `LookupExpr` specifies the
filter conditions that will be used to construct the spans for each lookup.

Currently, the only filter conditions that are allowed are equality and
`IN` expressions, representing constant filters and equi-join conditions.
The advantage of using an expression, however, is that it is easy to
extend in the future if we want to support other expressions (e.g. like
`OR` to support a disjunction on input columns).

Informs #59649

Release note (performance improvement): Added support for left and anti
lookup joins in cases where the equi-join condition is not on the first
column in the index, but the first column(s) are constrained to a small
number of constant values using a CHECK constraint or an ENUM type. Planning
a lookup join for these cases can significantly improve performance if
the left input to the join is much smaller than the right-hand side.

rytaft committed:
sql,opt: add support for lookup joins with multiple spans

This commit adds support for lookup joins that scan multiple spans in
the index per input row. This is needed for cases when a prefix of the
index columns are constrained to multiple constant values, such as the
values in a CHECK constraint or an ENUM type.

This is implemented by adding a new field to the LookupJoinPrivate in
the optimizer (and the lookupJoiner in the execution engine). The new
field is called LookupExpr, and similar to KeyCols in the LookupJoinPrivate
(lookupCols in the lookupJoiner), LookupExpr represents the part of the
join condition used to perform the lookup into the index. It should only
be set when KeyCols is empty. LookupExpr is used instead of KeyCols when
the lookup condition is more complicated than a simple equality between
input columns and index columns. In this case, LookupExpr specifies the
filter conditions that will be used to construct the spans for each lookup.

Currently, the only filter conditions that are allowed are equality and
IN expressions, representing constant filters and equi-join conditions.
The advantage of using an expression, however, is that it is easy to
extend in the future if we want to support other expressions (e.g. like
OR to support a disjunction on input columns).

Informs #59649

Release note (performance improvement): Added support for left and anti
lookup joins in cases where the equi-join condition is not on the first
column in the index, but the first column(s) are constrained to a small
number of constant values using a CHECK constraint or an ENUM type. Planning
a lookup join for these cases can significantly improve performance if
the left input to the join is much smaller than the right-hand side.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60302)
<!-- Reviewable:end -->

andy-kimball commented:
Fixes #51576

rytaft commented:
Thanks for pointing out that issue -- I wasn't aware of those use cases, and as written, this PR won't fix them. But I have an idea about how to improve this PR that would fix that issue. I'm going to give it a try and see if it works.

Please hold off on reviewing for now.





<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
I took an early high-level look. This is really cool!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-:-MTCaMix8Mjza6ItrWdv:b-36ekfo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @mgartner, and @rytaft)

___
*[pkg/sql/execinfrapb/processors_sql.proto, line 276 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTCRGDWDcULZKa_ma-5:-MTCRGDWDcULZKa_ma-6:b-g79jz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c8af9f1eca428f1881e0acd122fbf375bcf0345/pkg/sql/execinfrapb/processors_sql.proto#L276)):*
> ```protobuf
>   optional uint32 index_idx = 2 [(gogoproto.nullable) = false];
> 
>   // LookupColumns represents the part of the join condition used to perform
> ```

LookupExpr is strictly more general than LookupColumns, correct? It's worth mentioning that having lookup columns (a,b,c) is equivalent to having LookupExpr `a=idx-col-1 AND @2=idx-col-2 AND @3=idx-col-3` (or something like that?)

I'd also consider moving it after LookupExpr, and document that as just a more efficient way to represent the common case.
___
*[pkg/sql/execinfrapb/processors_sql.proto, line 301 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTCRB549Mdj0-mWsEQ9:-MTCRB549Mdj0-mWsEQA:bvcdjsg) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c8af9f1eca428f1881e0acd122fbf375bcf0345/pkg/sql/execinfrapb/processors_sql.proto#L301)):*
> ```protobuf
>   // more complicated than a simple equality between input columns and index
>   // columns. In this case, LookupExpr specifies the expression that will be
>   // used to construct the spans for each lookup.
> ```

We need to specify here exactly what expressions are supported.
___
*[pkg/sql/opt/ops/relational.opt, line 317 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTCahTU-ymqXUwYsEWz:-MTCahTU-ymqXUwYsEX-:bpffdlx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c8af9f1eca428f1881e0acd122fbf375bcf0345/pkg/sql/opt/ops/relational.opt#L317)):*
> ```
>     # The key columns are listed in the same order as the index columns (or a
>     # prefix of them).
>     KeyCols ColList
> ```

Do you think we should keep both? Is there a big downside to only using LookupExpr (and detecting the common case in the execbuilder)? (if not, can add a TODO for now)
___
*[pkg/sql/opt/ops/relational.opt, line 325 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTCbWQK1S5Lg2Kqqw1e:-MTCbWQK1S5Lg2Kqqw1f:baiu3wt) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c8af9f1eca428f1881e0acd122fbf375bcf0345/pkg/sql/opt/ops/relational.opt#L325)):*
> ```
>     # columns. In this case, LookupExpr specifies the filter conditions that
>     # will be used to construct the spans for each lookup.
>     LookupExpr FiltersExpr
> ```

It's strange that we point to a FiltersExpr in the private (it won't show up as a "child" of the rel expression). I guess that it works in practice since the expression is restricted (and we already have ConstFilters below), but we should explain here that it is restricted (e.g. it can't contain subqueries, or non-immutable expressions).
___
*[pkg/sql/rowexec/joinreader.go, line 106 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTCVOQbCpwrYGjE2BUV:-MTCVOQbCpwrYGjE2BUW:bxfny4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c8af9f1eca428f1881e0acd122fbf375bcf0345/pkg/sql/rowexec/joinreader.go#L106)):*
> ```Go
> 	// simple equality between input columns and index columns. In this case,
> 	// lookupExpr specifies the expression that will be used to construct the
> 	// spans for each lookup.
> ```

[nit] Should point to the spec for documentation on what kind of expressions are supported
___
*[pkg/sql/rowexec/joinreader_strategies.go, line 123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTCZvsG-wdBU8nXqUx0:-MTCZvsHBFozcEaTXy7z:b-o1a6pz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c8af9f1eca428f1881e0acd122fbf375bcf0345/pkg/sql/rowexec/joinreader_strategies.go#L123)):*
> ```Go
> 	// getLookupKeyColsCount returns the number of key columns used to lookup into
> 	// the index.
> 	getLookupKeyColsCount() int
> ```

This method feels strange, it's not clear conceptually why it would be necessary and what are the consequences of this value. Looks like it's only used to call `PartialKey` in `performLookup`; maybe we could move that call inside `processLookedUpRow` and let each implementation do what it needs to. I think this would also clean up the special casing for not calling `PartialKey` for `indexJoinReaderType`.

I'm thinking in particular of how things will work when we support more cases where the spans might not have the same number of keys. Example: index on `(a,b)` and ON condition `a=x AND b>=10`. Then each lookup would be of the form `[/x/10 - /(x+1))` and it's not clear what `getLookupKeyColsCount()` would be then.


<!-- Sent from Reviewable.io -->

rytaft commented:
Ok, I went down a pretty deep rabbit hole, and unfortunately I think supporting the use cases in #51576 would require a much bigger change than we have time for right now. 

For future reference, my idea was to represent the lookup spans as constraints with placeholders for the input columns. The optimizer would build these constraints using the `idxconstraints` library, and the `lookupJoiner` would fill in the placeholders with values for each input row. To implement this idea, inside `GenerateLookupJoins`, we would need to replace all the input columns in the filters with placeholders, and then use those filters to build index constraints. The biggest blocker with this idea is that we would need add support for placeholders in the `idxconstraints` library. This is something we have been wanting to do anyway, but as I discovered today, is nontrivial. There is a lot of code in `idxconstraints` that relies on the datums being comparable, and the only thing we can do with placeholders is check for equality. I do think this change is worth doing at some point, but unfortunately I don't have time right now (given that stability period is less than two weeks away).

So I think we should go with my current implementation. Thanks for already taking a look, @raduberinde, and for everyone else, this is ready for review.





<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I agree we should not try to implement the other use cases right now. But I don't see why the current approach wouldn't work. The joiner could separate the conditions in the lookupexpr into ones that involve columns before the equality and those that involve columns after the equality. The "before" expressions are only allowed to lead to single-key spans. The "after" expressions can lead to any spans. Each lookup uses the cartesian product between the before and after spans.

rytaft commented:
Yea, you're probably right. My thought was that there is already so much logic to handle these different scenarios inside `idxconstraint`, that it would be nice to avoid reinventing the wheel -- especially since we want to support placeholders there anyway. But maybe it's just too much effort for insufficient reward, since we could probably support most of the different use cases in #51576 with the `LookupExpr` approach. Anyway, I think we can punt on that decision for now...





<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60302)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @mgartner, and @RaduBerinde)

___
*[pkg/sql/execinfrapb/processors_sql.proto, line 276 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTCRGDWDcULZKa_ma-5:-MTCq4kz3J-S5Wi-qfZt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c8af9f1eca428f1881e0acd122fbf375bcf0345/pkg/sql/execinfrapb/processors_sql.proto#L276)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

LookupExpr is strictly more general than LookupColumns, correct? It's worth mentioning that having lookup columns (a,b,c) is equivalent to having LookupExpr `a=idx-col-1 AND @2=idx-col-2 AND @3=idx-col-3` (or something like that?)

I'd also consider moving it after LookupExpr, and document that as just a more efficient way to represent the common case.
</blockquote></details>

Done.
___
*[pkg/sql/execinfrapb/processors_sql.proto, line 301 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTCRB549Mdj0-mWsEQ9:-MTCq6_D14VDAZpNIAr2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c8af9f1eca428f1881e0acd122fbf375bcf0345/pkg/sql/execinfrapb/processors_sql.proto#L301)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We need to specify here exactly what expressions are supported.
</blockquote></details>

Done.
___
*[pkg/sql/opt/ops/relational.opt, line 317 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTCahTU-ymqXUwYsEWz:-MTD6exf2PLAMrYXM99M:b-7efhbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c8af9f1eca428f1881e0acd122fbf375bcf0345/pkg/sql/opt/ops/relational.opt#L317)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Do you think we should keep both? Is there a big downside to only using LookupExpr (and detecting the common case in the execbuilder)? (if not, can add a TODO for now)
</blockquote></details>

I don't think there's a big downside. It would make the execbuilder more complicated, but it would probably simplify the code a bit in `GenerateLookupJoins`. Added a TODO.
___
*[pkg/sql/opt/ops/relational.opt, line 325 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTCbWQK1S5Lg2Kqqw1e:-MTD8JBi7MQLTrNMqVKW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c8af9f1eca428f1881e0acd122fbf375bcf0345/pkg/sql/opt/ops/relational.opt#L325)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It's strange that we point to a FiltersExpr in the private (it won't show up as a "child" of the rel expression). I guess that it works in practice since the expression is restricted (and we already have ConstFilters below), but we should explain here that it is restricted (e.g. it can't contain subqueries, or non-immutable expressions).
</blockquote></details>

Done.
___
*[pkg/sql/rowexec/joinreader.go, line 106 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTCVOQbCpwrYGjE2BUV:-MTD8gwOELlUDwWeXSzj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c8af9f1eca428f1881e0acd122fbf375bcf0345/pkg/sql/rowexec/joinreader.go#L106)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] Should point to the spec for documentation on what kind of expressions are supported
</blockquote></details>

Done.
___
*[pkg/sql/rowexec/joinreader_strategies.go, line 123 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTCZvsG-wdBU8nXqUx0:-MTDBOWe9obhhH5Mi7FK:b-jd3rls) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c8af9f1eca428f1881e0acd122fbf375bcf0345/pkg/sql/rowexec/joinreader_strategies.go#L123)):*
I don't think it works to move the call inside `processLookedUpRow` since by that point we've already called `NextRow` and the key would be incorrect. We also don't have access to the fetcher (unless we pass it in).

> I'm thinking in particular of how things will work when we support more cases where the spans might not have the same number of keys

In order to make that work, we would need the key with the maximum number of columns in any of the spans. So maybe `getMaxLookupKeyCols()`?  (Although we'd also need to change some things about the implementation of the span generator to support this case since `getMatchingRowIndices` wouldn't be able to use the stringified key as the key of the map...) 

Anyway, I changed the name to `getMaxLookupKeyCols()`, but let me know if you have a better idea.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
execution changes <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> maybe @yuzefovich wants to take a look as well

> Reviewed 27 of 33 files at r1, 6 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-:-MTGHS8lCXaQNt31St0E:b-ub1z0b)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/rowexec/joinreader_span_generator.go, line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTGEsVU3gdaoDebQuII:-MTGEsVU3gdaoDebQuIJ:bnx2lmk) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e922ed92a23aab8b42a5227569808cfdaef527b/pkg/sql/rowexec/joinreader_span_generator.go#L131)):*
> ```Go
> 	// The optimizer should have ensured that all conditions fall into one of
> 	// these two categories. Any other expression types will return an error.
> 	var fillInIndexColInfos func(expr tree.TypedExpr) error
> ```

nit: can we extract this nested function easily? Might be cleaner
___
*[pkg/sql/rowexec/joinreader_strategies.go, line 75 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTGGThv1qH8-N0qP0p9:-MTGGThv1qH8-N0qP0pA:b-3bm37f) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e922ed92a23aab8b42a5227569808cfdaef527b/pkg/sql/rowexec/joinreader_strategies.go#L75)):*
> ```Go
> 
> // generateSpans is part of the joinReaderSpanGenerator interface.
> func (g *defaultSpanGenerator) generateSpans(rows []rowenc.EncDatumRow) (roachpb.Spans, error) {
> ```

nit: maybe move this to the same file as `multiSpanGenerator`


<!-- Sent from Reviewable.io -->

rytaft committed:
rowexec: move defaultSpanGenerator to joinreader_span_generator.go

Move the defaultSpanGenerator to joinreader_span_generator.go to consolidate
all implementations of joinReaderSpanGenerator in the same file.

Release note: None

rytaft reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-:-MTI4SCK456I0_CXu7HS:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @mgartner, and @RaduBerinde)

___
*[pkg/sql/rowexec/joinreader_span_generator.go, line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTGEsVU3gdaoDebQuII:-MTI3HjO72W01d8jgbQB:b-f1uqoq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e922ed92a23aab8b42a5227569808cfdaef527b/pkg/sql/rowexec/joinreader_span_generator.go#L131)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: can we extract this nested function easily? Might be cleaner
</blockquote></details>

Done. I needed to add a few of these local variables to the `multiSpanGenerator`, but I don't think it's too bad.
___
*[pkg/sql/rowexec/joinreader_strategies.go, line 75 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTGGThv1qH8-N0qP0p9:-MTI4Lpy5IaP1F5v-V42:bjff8zf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e922ed92a23aab8b42a5227569808cfdaef527b/pkg/sql/rowexec/joinreader_strategies.go#L75)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

nit: maybe move this to the same file as `multiSpanGenerator`
</blockquote></details>

Done. Did it in a new commit for clarity.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
What a doozy. I'll probably want to take another look after some rest, but looks good overall. I left a few comments.

> Reviewed 25 of 33 files at r1, 3 of 6 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-:-MTJ74pyDGNWcvwJkCT1:bcbwm4u)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt_exec_factory.go, line 615 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTIz7jA3NVeD7ksJ8UK:-MTIz7jA3NVeD7ksJ8UL:b-qtvmuq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db8d22d36d8033cea1b76bf6628492474b54815/pkg/sql/opt_exec_factory.go#L615)):*
> ```Go
> 	}
> 	pred := makePredicate(joinType, planColumns(input.(planNode)), planColumns(tableScan))
> 	if lookupExpr != nil && lookupExpr != tree.DBoolTrue {
> ```

When would `lookupExpr == tree.DBoolTrue`?
___
*[pkg/sql/opt/xform/join_funcs.go, line 309 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTJ54F-Cv5anP_FymMO:-MTJ54F-Cv5anP_FymMP:b-xz4y59) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db8d22d36d8033cea1b76bf6628492474b54815/pkg/sql/opt/xform/join_funcs.go#L309)):*
> ```Go
> 				lookupJoin.KeyCols = opt.ColList{}
> 				rightSideCols = opt.ColList{}
> 				break
> ```

It's confusing to me that this is inside the for-loop that is trying to find a way to constrain each column. Is there a sane way to move this after the for-loop?
___
*[pkg/sql/opt/xform/join_funcs.go, line 509 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTJ7VLQBwFVvbioJhfP:-MTJ7VLQBwFVvbioJhfQ:btaztis) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db8d22d36d8033cea1b76bf6628492474b54815/pkg/sql/opt/xform/join_funcs.go#L509)):*
> ```Go
> 
> 	eqFilters = make(memo.FiltersExpr, 0, len(filters))
> 	constFilters = make(memo.FiltersExpr, 0, len(filters))
> ```

Should this be removed? Looks like you're lazily initializing `constFilters` below in the `for` loop.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
TFTR! Updated.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-:-MTLhPq_CZHleyG8DcOE:b-p0fj95)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @mgartner, and @RaduBerinde)

___
*[pkg/sql/opt_exec_factory.go, line 615 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTIz7jA3NVeD7ksJ8UK:-MTLdlI_1UiUkMDzY_Zn:bl0m80n) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db8d22d36d8033cea1b76bf6628492474b54815/pkg/sql/opt_exec_factory.go#L615)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

When would `lookupExpr == tree.DBoolTrue`?
</blockquote></details>

Shouldn't ever be the case. Removed.
___
*[pkg/sql/opt/xform/join_funcs.go, line 309 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTJ54F-Cv5anP_FymMO:-MTLgkty0yhyZAWZwbv-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db8d22d36d8033cea1b76bf6628492474b54815/pkg/sql/opt/xform/join_funcs.go#L309)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

It's confusing to me that this is inside the for-loop that is trying to find a way to constrain each column. Is there a sane way to move this after the for-loop?
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/join_funcs.go, line 509 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTJ7VLQBwFVvbioJhfP:-MTLhDU_DGlS2E5MKRnp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0db8d22d36d8033cea1b76bf6628492474b54815/pkg/sql/opt/xform/join_funcs.go#L509)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Should this be removed? Looks like you're lazily initializing `constFilters` below in the `for` loop.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60302)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/execinfrapb/processors_sql.proto, line 297 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTMoB3S6E2MSDtNWpN3:-MTMoB3S6E2MSDtNWpN4:blb728c) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4b074b4975d8b6cda04e47ea8356275c0ebd097/pkg/sql/execinfrapb/processors_sql.proto#L297)):*
> ```protobuf
>   // For example, a valid LookupExpr for N=2 and M=2 might be:
>   //   @3 IN (10, 20) AND @2 = @4.
>   optional Expression lookup_expr = 16 [(gogoproto.nullable) = false];
> ```

I'm curious why the `lookup_columns` field is not being subsumed into this (here and in joinreader.go). Is it to avoid bumping up `MinAcceptedVersion` so that mixed version clusters can somewhat execute lookup/index join queries in a distributed manner (though IIUC that only works with an old gateway node, which seems a deficiency of the sql versioning scheme)?
And I didn't notice a change to `execinfra/version.go`.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60302)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @mgartner, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/exec/factory.opt, line 218 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTNC6MW2BwU0f0TReWw:-MTNC6MW2BwU0f0TReWx:br7vbu1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8e5f1e724bc1b8523a5718ee1b293480006f70c/pkg/sql/opt/exec/factory.opt#L218)):*
> ```
>     Index cat.Index
>     EqCols []exec.NodeColumnOrdinal
>     EqColsAreKey bool
> ```

EqColsAreKey needs some more explanation now.. Does it also cover the case where we use LookupCols and the spans are known to be keys?
___
*[pkg/sql/opt/ops/relational.opt, line 334 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTN8G0B-dojGBrjcWRB:-MTN8G0B-dojGBrjcWRC:buy3pfx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8e5f1e724bc1b8523a5718ee1b293480006f70c/pkg/sql/opt/ops/relational.opt#L334)):*
> ```
>     #  2. equalities between an index column and a constant, and
>     #  3. IN expressions between an index column and a tuple of constants.
>     LookupExpr FiltersExpr
> ```

[nit] Also add that all the index columns that show up in FiltersExpr must correspond to a contiguous prefix of the index columns.
___
*[pkg/sql/opt/xform/join_funcs.go, line 310 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTN9KDS4yd0l4iDrZZi:-MTN9KDS4yd0l4iDrZZj:b-9jova5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8e5f1e724bc1b8523a5718ee1b293480006f70c/pkg/sql/opt/xform/join_funcs.go#L310)):*
> ```Go
> 		}
> 
> 		if len(lookupJoin.KeyCols) == 0 && (joinType == opt.LeftJoinOp || joinType == opt.AntiJoinOp) {
> ```

What if we had two leading columns; the first one had one value and the second one had multiple values.. We'd still want to use the new infrastructure in that case (or we'd be doing what I like to call a "trash lookup join").

By the way, in the existing code, we will generate a lookup join even if we only matched constant values and there is no equality with an input column. This seems strictly worse than a hash join where one side is a constrained scan with those same filters.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60302)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @mgartner, @RaduBerinde, and @sumeerbhola)

___
*[pkg/sql/execinfrapb/processors_sql.proto, line 297 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTMoB3S6E2MSDtNWpN3:-MTb0qDBBhVY5Y1hUMtL:b-fwfydy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4b074b4975d8b6cda04e47ea8356275c0ebd097/pkg/sql/execinfrapb/processors_sql.proto#L297)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I'm curious why the `lookup_columns` field is not being subsumed into this (here and in joinreader.go). Is it to avoid bumping up `MinAcceptedVersion` so that mixed version clusters can somewhat execute lookup/index join queries in a distributed manner (though IIUC that only works with an old gateway node, which seems a deficiency of the sql versioning scheme)?
And I didn't notice a change to `execinfra/version.go`.
</blockquote></details>

It was just to minimize the total number of changes required and potential impact on existing lookup joins. I think you're right that eventually we'll want to combine these two. Added a TODO.

And good point about the DistSQL version -- I updated `execinfra/version.go`.
___
*[pkg/sql/opt/exec/factory.opt, line 218 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTNC6MW2BwU0f0TReWw:-MTb5_ExDdvh4XCH0JqY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8e5f1e724bc1b8523a5718ee1b293480006f70c/pkg/sql/opt/exec/factory.opt#L218)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

EqColsAreKey needs some more explanation now.. Does it also cover the case where we use LookupCols and the spans are known to be keys?
</blockquote></details>

Done.
___
*[pkg/sql/opt/ops/relational.opt, line 334 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTN8G0B-dojGBrjcWRB:-MTb5rcK3UjrGUCWY0jD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8e5f1e724bc1b8523a5718ee1b293480006f70c/pkg/sql/opt/ops/relational.opt#L334)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] Also add that all the index columns that show up in FiltersExpr must correspond to a contiguous prefix of the index columns.
</blockquote></details>

Done.
___
*[pkg/sql/opt/xform/join_funcs.go, line 310 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTN9KDS4yd0l4iDrZZi:-MTbAFei1ghl_qEoir9T:b3jwyky) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8e5f1e724bc1b8523a5718ee1b293480006f70c/pkg/sql/opt/xform/join_funcs.go#L310)):*
> What if we had two leading columns; the first one had one value and the second one had multiple values.. We'd still want to use the new infrastructure in that case (or we'd be doing what I like to call a "trash lookup join").

Fixed and added a test.

> By the way, in the existing code, we will generate a lookup join even if we only matched constant values and there is no equality with an input column. This seems strictly worse than a hash join where one side is a constrained scan with those same filters.

I think the code prevents this. `findFiltersForIndexLookup` checks that there was at least one equality filter, and if not, returns nil for both slices. There's also a test for this case with `expect-not=GenerateLookupJoins` in the `join` test file. Or am I misunderstanding?


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 13 of 22 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-:-MTbWOdjCi6YgU9SksCW:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @mgartner, @RaduBerinde, and @sumeerbhola)

___
*[pkg/sql/opt/exec/factory.opt, line 211 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTbUL-o-M6uSfvNlMDd:-MTbUL-o-M6uSfvNlMDe:b-pnmeqp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff23122299eb8a6e9d6e2a01b582a6d447eda1b4/pkg/sql/opt/exec/factory.opt#L211)):*
> (eqColsAreKey will be true in this case if the columns that are part of the equi-join conditions form a key in the table)

If we add more complex expressions, like disjunctions, then this will have to be more sophiscated in order to preserve the meaning of `eqColsAreKey` as "each input row matches with at most one index row".

I think the disjunction `x IN (...)` that is currently supported is fine as long as x is a key and the RHS is constants.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-:-MTcPcev6RFv9ZxkSWDq:b-ucfknh)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @mgartner, @RaduBerinde, and @sumeerbhola)

___
*[pkg/sql/opt/exec/factory.opt, line 211 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTbUL-o-M6uSfvNlMDd:-MTcOaKDAaFMcP46jcK1:b-t66kuf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff23122299eb8a6e9d6e2a01b582a6d447eda1b4/pkg/sql/opt/exec/factory.opt#L211)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

> (eqColsAreKey will be true in this case if the columns that are part of the equi-join conditions form a key in the table)

If we add more complex expressions, like disjunctions, then this will have to be more sophiscated in order to preserve the meaning of `eqColsAreKey` as "each input row matches with at most one index row".

I think the disjunction `x IN (...)` that is currently supported is fine as long as x is a key and the RHS is constants.
</blockquote></details>

I think we should continue to determine the value of `eqColsAreKey` based only on the filters that are simple equalities (that's what I meant by equi-join conditions, but can rephrase if you think it's unclear). For example, if the input columns are `(a, b, c)` and the index columns are `(x, y, z)`, given the join condition `a = x AND (b = y OR c = z)`, `eqColsAreKey` should only be true if `x` is a key.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60302)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto, @mgartner, @RaduBerinde, and @sumeerbhola)

___
*[pkg/sql/opt/exec/factory.opt, line 211 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTbUL-o-M6uSfvNlMDd:-MTcVJ166701i_rF1Q9Q:bqc7ws2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff23122299eb8a6e9d6e2a01b582a6d447eda1b4/pkg/sql/opt/exec/factory.opt#L211)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think we should continue to determine the value of `eqColsAreKey` based only on the filters that are simple equalities (that's what I meant by equi-join conditions, but can rephrase if you think it's unclear). For example, if the input columns are `(a, b, c)` and the index columns are `(x, y, z)`, given the join condition `a = x AND (b = y OR c = z)`, `eqColsAreKey` should only be true if `x` is a key.
</blockquote></details>

I see. I didn't understand what you mean by "equi-join conditions". I thought it was synonymous with lookupExpr, which woudn't have the same guarantees, like your example shows.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60302)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @asubiotto, @mgartner, @RaduBerinde, and @sumeerbhola)

___
*[pkg/sql/opt/exec/factory.opt, line 211 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTbUL-o-M6uSfvNlMDd:-MTcZ7HK9WOZYrgsErjv:b6d7wzq) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff23122299eb8a6e9d6e2a01b582a6d447eda1b4/pkg/sql/opt/exec/factory.opt#L211)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I see. I didn't understand what you mean by "equi-join conditions". I thought it was synonymous with lookupExpr, which woudn't have the same guarantees, like your example shows.
</blockquote></details>

Updated the text to be clearer.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60302)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @asubiotto, @mgartner, @RaduBerinde, and @sumeerbhola)

___
*[pkg/sql/opt/exec/factory.opt, line 211 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-MTbUL-o-M6uSfvNlMDd:-MTcdvZzBnYPCPNxcl-t:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff23122299eb8a6e9d6e2a01b582a6d447eda1b4/pkg/sql/opt/exec/factory.opt#L211)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Updated the text to be clearer.
</blockquote></details>

Thanks!


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice work!

> Reviewed 12 of 33 files at r1, 19 of 22 files at r7, 1 of 1 files at r8, 2 of 2 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-:-MTgVAJ438PK9B2m5d9d:bcmw7ss)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @sumeerbhola)



<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Thanks all!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60302#-:-MTgWgUl5FG6i8xjJeDB:bbdmny1)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @sumeerbhola)



<!-- Sent from Reviewable.io -->
